server/lib/src/idm/identityverification.rs

```diff
@@ -0,0 +1,641 @@
+        } else {
+            IdentifyUserResponse::Success
+        };
+        Ok(res)
+    }
+
+    // End of public functions
+
+    fn check_for_early_return_conditions(
+        &mut self,
+        ident: &Identity,
+        target: &Uuid,
+    ) -> Result<Option<IdentifyUserResponse>, OperationError> {
+        // here we check that the identify user feature is available before we do anything else
+        if !self.check_if_identify_feature_available(ident)? {
+            return Ok(Some(IdentifyUserResponse::IdentityVerificationUnavailable));
+        };
+
+        if !self.is_valid_user_uuid(ident, target)? {
+            return Ok(Some(IdentifyUserResponse::InvalidUserId));
+        };
+        // here we check if the user provided their own uuid, if they did we just respond with IdentityVerificationAvailable.
+        if ident.get_uuid().eq(&Some(*target)) {
+            return Ok(Some(IdentifyUserResponse::IdentityVerificationAvailable));
+        };
+        Ok(None)
+    }
+
+    fn check_if_identify_feature_available(
+        &mut self,
+        ident: &Identity,
+    ) -> Result<bool, OperationError> {
+        let search = match SearchEvent::from_whoami_request(ident.clone(), &self.qs_read) {
+            Ok(s) => s,
+            Err(e) => {
+                admin_error!("Failed to generate whoami search event: {:?}", e);
+                return Err(e);
+            }
+        };
+        self.qs_read
+            .search(&search)
+            .and_then(|mut entries| entries.pop().ok_or(OperationError::NoMatchingEntries))
+            .map(
+                |entry| match entry.get_ava_single_eckey_private("id_verification_eckey") {
+                    Some(key) => key.check_key().is_ok(),
+                    None => false,
+                },
+            )
+    }
+
+    fn is_valid_user_uuid(
+        &mut self,
+        ident: &Identity,
+        target: &Uuid,
+    ) -> Result<bool, OperationError> {
+        let search =
+            match SearchEvent::from_target_uuid_request(ident.clone(), *target, &self.qs_read) {
+                Ok(s) => s,
+                Err(e) => {
+                    admin_error!("Failed to retrieve user with the given UUID: {:?}", e);
+                    return Err(e);
+                }
+            };
+
+        let user_entry = self
+            .qs_read
+            .search(&search)
+            .and_then(|mut entries| entries.pop().ok_or(OperationError::NoMatchingEntries))?;
+
+        match user_entry.get_ava_single_eckey_public("id_verification_eckey") {
+            Some(key) => Ok(key.check_key().is_ok()),
+            None => Ok(false),
+        }
+    }
+
+    fn get_user_own_key(&mut self, ident: &Identity) -> Result<EcKey<Private>, OperationError> {
+        let search = match SearchEvent::from_whoami_request(ident.clone(), &self.qs_read) {
+            Ok(s) => s,
+            Err(e) => {
+                admin_error!("Failed to retrieve user with the given UUID: {:?}", e);
```

Comment:
The UUID should be included in the log, especially since it says "with the given UUID" 😄

Reply:
sure :)